components/brave_new_tab_ui/containers/newTab/gridTile.tsx

```diff
@@ -3,158 +3,151 @@
+  const { site, isMenuShowing, children, draggable } = props
+  const style = useMemo(() => ({
+    transform: CSS.Transform.toString(draggable?.transform || null),
+    transition: draggable?.transition,
+    opacity: draggable?.isDragging ? 0 : 1
+  }), [draggable])
+
+  return <Tile
+    ref={draggable?.setNodeRef} {...draggable?.attributes} {...draggable?.listeners}
+    isDragging={!!draggable?.isDragging}
+    isMenuShowing={!!isMenuShowing}
+    title={site.title}
+    href={site.url}
+    style={style}>
+    {children}
+    <TileFavicon src={generateGridSiteFavicon(site)} />
+    <TileTitle>{site.title}</TileTitle>
+  </Tile>
+}
 
-  handleMoveOutside = (event: Event) => {
-    if (
-      this.tileMenuRef &&
-      this.tileMenuRef.current &&
-      !this.tileMenuRef.current.contains(event.target)
-    ) {
-      this.setState({ showMenu: false })
+function TopSite (props: Props) {
+  const { siteData, disabled } = props
+
+  const tileMenuRef = useRef<any>()
+  const sortable = useSortable({ id: siteData.id, disabled })
+  const [showMenu, setShowMenu] = useState(false)
+
+  const handleClickOutside = useCallback((e: Event) => {
+    if (!tileMenuRef.current || tileMenuRef.current.contains(e.target)) { return }
+    setShowMenu(false)
+  }, [])
+
+  const handleMoveOutside = useCallback((e: Event) => {
+    if (!tileMenuRef.current || tileMenuRef.current.contains(e.target)) { return }
+    setShowMenu(false)
+  }, [])
+
+  const onKeyPressSettings = useCallback((e: KeyboardEvent) => {
+    if (e.key === 'Escape') { setShowMenu(false) }
+  }, [])
+
+  useEffect(() => {
+    document.addEventListener('mousedown', handleClickOutside)
+    document.addEventListener('mouseMove', handleMoveOutside)
+    document.addEventListener('keydown', onKeyPressSettings)
+    return () => {
+      document.removeEventListener('mousedown', handleClickOutside)
+      document.removeEventListener('mousemove', handleMoveOutside)
+      document.removeEventListener('keydown', onKeyPressSettings)
     }
-  }
+  }, [handleClickOutside, handleMoveOutside, onKeyPressSettings])
 
-  onIgnoredTopSite (site: NewTab.Site, e: Event) {
+  const onShowTileMenu = useCallback((e: React.MouseEvent) => {
     e.preventDefault()
-    this.setState({ showMenu: false })
-    this.props.actions.tileRemoved(site.url)
-  }
+    setShowMenu(true)
+  }, [])
 
-  onEditTopSite (site: NewTab.Site, e: Event) {
+  const onIgnoredTopSite = useCallback((site: NewTab.Site, e: React.MouseEvent) => {
     e.preventDefault()
-    this.setState({ showMenu: false })
-    this.props.onShowEditTopSite(site)
-  }
+    setShowMenu(false)
+    props.actions.tileRemoved(site.url)
+  }, [props.actions.tileRemoved])
 
-  onShowTileMenu = (e: React.MouseEvent<HTMLButtonElement, MouseEvent>) => {
+  const onEditTopSite = useCallback((site: NewTab.Site, e: React.MouseEvent) => {
```

Comment:
I think we've gone back and forth on the usefullness of the `React.useCallback` optimization vs just having the functions be in-line and re-created every render. I think where we're currently at is that it's an over-optimization and may do a tiny bit more harm than good, especially if the memoized function isn't being sent to a memoized component. In other words, if useCallback doesn't prevent a child from re-rendering, then perhaps we should just use regular non-useCallback'ed functions everywhere. What do you think?

Reply:
Fine by me. I mostly do it without thinking now but it was definitely a learned habit. Happy to unlearn because it definitely makes things a bit simpler.

Reply:
I also kind of like having the event handlers right where they're used :)